### PR TITLE
Cleaning up HAL buffer transfer usage.

### PR DESCRIPTION
**Pull request by benvanik**
This cleans up places using the `iree_hal_buffer_*` API in an effort to move more toward device-based transfers. In order to make the requirement of the buffer functions acting on mappable buffers more explicit "map" was added to their name.

### CODE REVIEW

**Review by ScottTodd — APPROVED** [review]
Thanks!

**Comment by benvanik** [comment]
Failing test was the asan flake in scalar.mlir.

File: iree/hal/cuda/cuda_allocator.c [hunk 1]
```diff
@@ -220,7 +222,7 @@ static iree_status_t iree_hal_cuda_allocator_allocate_buffer(
   iree_hal_buffer_t* buffer = NULL;
   if (iree_status_is_ok(status)) {
     status = iree_hal_cuda_buffer_wrap(
-        (iree_hal_allocator_t*)allocator, memory_type, params->access,
+        allocator->base_device, base_allocator, memory_type, params->access,
         params->usage, allocation_size,
         /*byte_offset=*/0,
         /*byte_length=*/allocation_size, device_ptr, host_ptr, &buffer);
```
Comment (ScottTodd) [inline iree/hal/cuda/cuda_allocator.c:228]: compile error caught on postsubmit:
```
../iree/hal/cuda/cuda_allocator.c:228:64: error: too many arguments to function call, expected 10, have 11
        /*byte_length=*/allocation_size, device_ptr, host_ptr, &buffer);
                                                               ^~~~~~~
../iree/hal/cuda/cuda_buffer.h:19:1: note: 'iree_hal_cuda_buffer_wrap' declared here
iree_status_t iree_hal_cuda_buffer_wrap(
^
```

Comment (hanhanW) [reply iree/hal/cuda/cuda_allocator.c:228]: I just hit the same issue when building IREE with dbg mode.

Comment (ScottTodd) [reply iree/hal/cuda/cuda_allocator.c:228]: Sent a proposed fix: https://github.com/google/iree/pull/8566

**Pull request merged**